Add ability to escape {{EXPRESSION}} syntax from #75

JonathanSteininger opened the pull request:
This is to solve this [issue](https://github.com/hyprwm/hyprlang/issues/75). On the 0.6.3 hyprlang release, the ability to do simple arithmetics were added with the {{VAL1 <+,-,*,/> VAL2}} syntax. 
The only issue is that this is un-escapable, so in simple binds or commands that use this syntax, most notably the playerctl's formatted output.

To add the ability to escape this syntax, I simply added the ability to add '\' to multiple parts of the expression.
`\{{   {\{   \{\{` will all escape from the arithmetic expression. This also works for "}" but theres no point in doing so.

I have tested this on my main configuration and on a simple config with this:
`exec=hyprctl notify 5000 50000 'rgb(ffffff)' "{{1 + 2}} {\{test}} {\{20 - 10}} \{\{10 + 10}} \}} \}\} {{ \{{ \\{{"`
with this output in the message:
`3 {{test}} {{20 - 10}} {{10 + 10}} }} }} {{ {{ \{{`

vaxerski commented on src/config.cpp:
```diff
@@ -642,10 +642,23 @@ CParseResult CConfig::parseLine(std::string line, bool dynamic) {
-                const auto BEGIN_EXPR = RHS.find("{{");
+                auto firstUnescaped = RHS.find("{{");
+                //keep searching until the escape char is not found.
+                while(RHS.at(firstUnescaped - 1) == '\\'){
```
uflow if firstUnescaped == 0

vaxerski replied:
we all make mistakes

vaxerski commented on src/config.cpp:
```diff
@@ -642,10 +642,23 @@ CParseResult CConfig::parseLine(std::string line, bool dynamic) {
+                        break;
+                }
+                //real match was never found.
+                if(firstUnescaped == std::string::npos){
```
no {}

vaxerski reviewed (commented):
also missing format

vaxerski reviewed (commented):
c-f needed, also please add tests.

JonathanSteininger commented:
I have added a bunch of test which I think covers all cases where this would be used.

I rewrote a bunch of it to be able to escape escapes. Since in the previous version it was impossible to have `\` as text before an expression. now it can be escaped with `\\`

before: `\\{{10 + 10}}` -> `\{{10 + 10}}`
now: `\\{{10 + 10}}` -> `\20`

JonathanSteininger commented:
Also, I rewrote the deletion of escapes to handle deleting escaped escapes better.

Had to move it below processing the recursive variables, so they are escaped while processing vars aswell. oops.

JonathanSteininger commented:
> c-f needed, also please add tests.

Just asking, when you were referring to c-f. It meant code-formatting right?

JonathanSteininger commented:
> `clang-format -i src/**/*{cpp,hpp}` also should prolly `clang-format -i tests/**/*{cpp,hpp}`

Thanks, Ill remember from now on.

JonathanSteininger commented:
Added documentation pull request for it as well.

vaxerski commented on src/config.cpp:
```diff
@@ -672,6 +696,27 @@ CParseResult CConfig::parseLine(std::string line, bool dynamic) {
+                continue;
+            }
+            //checks if any of the chars were escapable.
+            for (unsigned int j = 0; j < sizeof(ESCAPE_SET) / sizeof(char); j++) {
```
this is very C. Instead, make escape_set a `std::array` and use a foreach loop `for (const auto& c : escape_set)`

vaxerski reviewed (commented):
rest lgtm, pretty nice

JonathanSteininger commented:
This is c-f'd already and still passes all the tests.

vaxerski reviewed (approved):
thanks!

JonathanSteininger commented:
I am sorry that I have to ask, this is the first pull request I have made on an external repo.

But what are the next steps? Is everything done and only your pull is needed? or do I still need to do something?

vaxerski commented:
I just need to press merge which I havent done yet because I had this tab open and was waiting for tests to pass and then forgot
